### ISSUE TITLE
How to check Association rules has how many hits on new records?

### ISSUE THREAD

**Issue by viciaky** [rasbt/mlxtend#343]
As the issue title, i split the transaction records into train and test, the train set derived association rules. How can i check how many records in test set have been hit by these rules? Any function in the package i can use?

**Comment by rasbt** [comment]
Since association rule mining would be more like a "unsupervised" learning task, there's currently no API for the separate handling of training and testing. However, looking up entries in a second dataset (or test dataset) that have been covered by the rule should be possible. It's a bit clunky at the moment now. 

Let's say you have generated the following rules from a training dataset

## Generate rules


```python

import pandas as pd
from mlxtend.preprocessing import OnehotTransactions
from mlxtend.frequent_patterns import apriori


dataset = [['Milk', 'Onion', 'Nutmeg', 'Kidney Beans', 'Eggs', 'Yogurt'],
           ['Dill', 'Onion', 'Nutmeg', 'Kidney Beans', 'Eggs', 'Yogurt'],
           ['Milk', 'Apple', 'Kidney Beans', 'Eggs'],
           ['Milk', 'Unicorn', 'Corn', 'Kidney Beans', 'Yogurt'],
           ['Corn', 'Onion', 'Onion', 'Kidney Beans', 'Ice cream', 'Eggs']]

oht = OnehotTransactions()
oht_ary = oht.fit(dataset).transform(dataset)
df_train = pd.DataFrame(oht_ary, columns=oht.columns_)

frequent_itemsets_train = apriori(df_train, min_support=0.6, use_colnames=True)

frequent_itemsets_train
```

![screen shot 2018-03-14 at 9 23 11 am](https://user-images.githubusercontent.com/5618407/37404830-637ba326-2769-11e8-9a8d-6e3bfd07a6d9.png)


```python
from mlxtend.frequent_patterns import association_rules

training_rules = association_rules(frequent_itemsets, metric="confidence", min_threshold=0.7)

training_rules
```

![screen shot 2018-03-14 at 9 23 50 am](https://user-images.githubusercontent.com/5618407/37404864-725d966a-2769-11e8-8d55-d2170b06dbca.png)


## Selecting rules

No, let's say we are interested in looking at the rule in the 2nd row (index position 1):

```python
training_rules.values[1]
array([frozenset({'Eggs'}), frozenset({'Kidney Beans'}), 0.8, 1.0, 1.0], dtype=object)
```

Note that these are frozen-sets to make the algorithm more efficient during evaluation. To join the antecedent (first item) and consequent (second item) in a set, we can do sth as follows:

```python
# join sets
itemset = set(training_rules.values[1][0]) | set(training_rules.values[1][1]) 
itemset
{'Eggs', 'Kidney Beans'}
```

## Looking up the itemset of interest in the test set

Next, let's assume we have a test set that is formatted similar to the `frequent_itemsets_train` set earlier. You may notice that the frequent itemsets are lists in these arrays (this could maybe be changed to frozensets in future for efficiency). Thus, two look up the itemset from the previous step in the itemsets contained in the data frame, we need to convert the itemsets from list to set representations first:

```
frequent_itemsets_test['itemsets'] = frequent_itemsets_test['itemsets'].apply(lambda x: set(x))
frequent_itemsets_test
```

![screen shot 2018-03-14 at 9 31 36 am](https://user-images.githubusercontent.com/5618407/37405243-8472a7c2-276a-11e8-837e-c87bc8014c73.png)

Finally, we can identify the itemset that matches our rule as follows:

```
frequent_itemsets_test[frequent_itemsets_test['itemsets'] == itemset]
```

![screen shot 2018-03-14 at 9 33 32 am](https://user-images.githubusercontent.com/5618407/37405337-c9cca9da-276a-11e8-9ebb-65ba42d43ff9.png)

Since the support for this itemset is 0.6, it means that 60% of the data instances in the test set match this pattern. To compare training and test sets, you want to see, ideally, that the support for training and test set are the same, I'd say.

I hope this helps!

(I am going to leave this issue open right now, since I think a cleaned-up version of this would be worthwhile adding to the documentation in future)

**Comment by viciaky** [comment]
Thank you very much~! Yesterday i was trying to do something like you said. This really help me a lot. 👍

**Comment by lucatoldo** [comment]
Very useful indeed.

**Comment by Tanishk-Sharma** [comment]
Thanks, @rasbt for instructions on how to look up the itemsets.

My issue:
Looking up antecedents and consequents in a association_rules dataframe was a headache even when using .astype('str').str.contains('(Something)') ... 
- it gives an error when there are multiple items, for example, (Foo, Bar)
- using regex=False didn't help either

the antecedends and consequents columns are type object, and association_rule['antecedents'].astype('str')  ---> converts them to frozenset({'Foo', 'Bar'})

What I did:
association_rule['antecedents'] = association_rule['antecedents'].apply(lambda x: set(x))
association_rule['consequents'] = association_rule['consequents'].apply(lambda x: set(x))

Is there a better way for accessing these now?

**Comment by rasbt** [comment]
Thanks for the feedback. The API is still unchanged and based on the frozensets; that's for efficiency reasons. You approach regarding `.apply(lambda x: set(x))` seems to be the most reasonable right now.
